File: data/harnesses/Plus/go/secure/CWE-77_10/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)
func zipDirectory(srcDir, outZip string) (string, error) {
	if srcDir == "" {
		return "", fmt.Errorf("src_dir must be a string")
	}
	if outZip == "" {
		return "", fmt.Errorf("out_zip must be a string")
	}
	if strings.TrimSpace(srcDir) == "" {
		return "", fmt.Errorf("src_dir must be non-empty")
	}

	srcInfo, err := os.Stat(srcDir)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("src_dir not found")
		}
		return "", fmt.Errorf("src_dir not found")
	}
	if !srcInfo.IsDir() {
		return "", fmt.Errorf("src_dir not found")
	}

	if !strings.HasSuffix(outZip, ".zip") {
		return "", fmt.Errorf("out_zip must end with .zip")
	}

	baseName := filepath.Base(outZip)
	if baseName != outZip || baseName == "." || baseName == ".." {
		return "", fmt.Errorf("out_zip must be a file name, not a path")
	}

	matched, err := regexp.MatchString(`^[A-Za-z0-9_.-]+\.zip$`, outZip)
	if err != nil {
		return "", fmt.Errorf("invalid out_zip name")
	}
	if !matched {
		return "", fmt.Errorf("invalid out_zip name")
	}

	cmd := exec.Command("zip", "-r", outZip, srcDir)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("zip failed: %s: %w", strings.TrimSpace(string(output)), err)
	}

	return fmt.Sprintf("Archive created: %s", outZip), nil
}

func assertRaises(fn func() (string, error)) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Panic during assertRaises: %v\n", r)
			os.Exit(1)
		}
	}()
	result, err := fn()
	if err == nil {
		fmt.Printf("Expected an error but none was raised. Got result: %s\n", result)
		os.Exit(1)
	}
}

func main() {
	base := filepath.Join("Test", "CWE-77_10")
	src := filepath.Join(base, "srcdir")

	// Clean up any previous test artifacts
	if _, err := os.Stat(base); err == nil {
		os.RemoveAll(base)
	}
	if err := os.MkdirAll(src, 0755); err != nil {
		fmt.Printf("Failed to create test directories: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(filepath.Join(src, "a.txt"), []byte("hello"), 0644); err != nil {
		fmt.Printf("Failed to write a.txt: %v\n", err)
		os.Exit(1)
	}
	if err := os.WriteFile(filepath.Join(src, "b.txt"), []byte("world"), 0644); err != nil {
		fmt.Printf("Failed to write b.txt: %v\n", err)
		os.Exit(1)
	}

	// Test normal archive name
	res, err := zipDirectory(src, "out.zip")
	if err != nil {
		// If zip tool is missing, it's acceptable behavior
		if !strings.Contains(err.Error(), "zip failed") {
			fmt.Printf("Unexpected error creating zip: %v\n", err)
			os.Exit(1)
		}
	} else {
		if res != "Archive created: out.zip" {
			fmt.Printf("Unexpected result: %s\n", res)
			os.Exit(1)
		}
		if _, statErr := os.Stat("out.zip"); os.IsNotExist(statErr) {
			fmt.Println("out.zip was not created")
			os.Exit(1)
		}
		os.Remove("out.zip")
	}

	// Test invalid src_dir
	assertRaises(func() (string, error) {
		return zipDirectory(filepath.Join(base, "missing"), "out.zip")
	})

	// Test invalid out_zip names / injection-like inputs
	assertRaises(func() (string, error) {
		return zipDirectory(src, "../out.zip")
	})
	assertRaises(func() (string, error) {
		return zipDirectory(src, "out; rm -rf /.zip")
	})
	assertRaises(func() (string, error) {
		return zipDirectory(src, "out.zip && whoami")
	})
	assertRaises(func() (string, error) {
		return zipDirectory(src, "out.tar")
	})

	// Test empty strings (equivalent to None/TypeError checks in Python)
	assertRaises(func() (string, error) {
		return zipDirectory("", "out.zip")
	})
	assertRaises(func() (string, error) {
		return zipDirectory(src, "")
	})

	// Clean up test directory
	os.RemoveAll(base)

	fmt.Println("All tests passed!")
}

File: data/harnesses/Plus/cpp/secure/CWE-77_10/main.cpp
#include <filesystem>
#include <string>
#include <regex>
#include <cstdlib>
#include <stdexcept>
#include <fstream>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

std::string zip_directory(const std::string& src_dir, const std::string& out_zip) {
    if (src_dir.empty()) {
        throw std::invalid_argument("src_dir must be non-empty");
    }
    
    bool is_src_dir_whitespace = true;
    for (char c : src_dir) {
        if (!std::isspace(static_cast<unsigned char>(c))) {
            is_src_dir_whitespace = false;
            break;
        }
    }
    if (is_src_dir_whitespace) {
        throw std::invalid_argument("src_dir must be non-empty");
    }

    if (!fs::is_directory(src_dir)) {
        throw std::runtime_error("src_dir not found");
    }

    if (out_zip.size() < 4 || out_zip.compare(out_zip.size() - 4, 4, ".zip") != 0) {
        throw std::invalid_argument("out_zip must end with .zip");
    }

    if (fs::path(out_zip).filename().string() != out_zip) {
        throw std::invalid_argument("out_zip must be a file name, not a path");
    }

    std::regex out_zip_regex(R"([A-Za-z0-9_.-]+\.zip)");
    if (!std::regex_match(out_zip, out_zip_regex)) {
        throw std::invalid_argument("invalid out_zip name");
    }

    std::string command = "zip -r " + out_zip + " " + src_dir;
    int ret = std::system(command.c_str());
    if (ret != 0) {
        throw std::runtime_error("zip failed");
    }

    return "Archive created: " + out_zip;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected '" + expected + "', got '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    }
}

int main() {
    fs::path base = fs::current_path() / "Test" / "CWE-77_10";
    fs::path src = base / "srcdir";

    // Clean up any previous test artifacts
    if (fs::exists(base)) {
        fs::remove_all(base);
    }
    fs::create_directories(src);

    {
        std::ofstream f(src / "a.txt");
        f << "hello";
    }
    {
        std::ofstream f(src / "b.txt");
        f << "world";
    }

    // Normal archive name (creation may depend on presence of system zip, but must not be injectable)
    // If zip is missing, RuntimeError is acceptable from the secure implementation.
    try {
        std::string res = zip_directory(src.string(), "out.zip");
        expect_equal(res, "Archive created: out.zip", "Normal zip_directory result mismatch");
        expect_true(fs::is_regular_file("out.zip"), "out.zip should exist");
        fs::remove("out.zip");
    } catch (const std::runtime_error&) {
        // zip tool may be absent; still acceptable behavior
    }

    // Invalid src_dir
    assert_raises([&]() {
        zip_directory((base / "missing").string(), "out.zip");
    }, "Missing src_dir should throw");

    // Invalid out_zip names / injection-like inputs
    assert_raises([&]() {
        zip_directory(src.string(), "../out.zip");
    }, "Path traversal out_zip should throw");

    assert_raises([&]() {
        zip_directory(src.string(), "out; rm -rf /.zip");
    }, "Semicolon injection out_zip should throw");

    assert_raises([&]() {
        zip_directory(src.string(), "out.zip && whoami");
    }, "Command chaining out_zip should throw");

    assert_raises([&]() {
        zip_directory(src.string(), "out.tar");
    }, "Non-zip extension out_zip should throw");

    // Cleanup
    fs::remove_all(base);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}